src/tools/tests/allocator-recipe-resolver-test.ts

```diff
@@ -1149,6 +1157,167 @@ recipe ReadingRecipeAD
+      ${schemas}
+
+      policy Policy {
+         @maxAge(age: '2d')
+         @allowedRetention(medium: 'Disk', encryption: false)
+         ${policy}
+      }
+      `;
+      await verifyWritingRecipeWithPolicy(recipe, policies, expected);
+    };
+
+    const thingSchema = `
+      schema Thing
+        a: Text
+        b: Text
+        c: Text
+        d: Text
+        e: Text
+    `;
+    await verifyPhantomReader(
+      thingSchema,
+      'from Thing access {a, b, d}',
+      'Thing',
+      'Thing {a: Text, b: Text, d: Text}'
+    );
+    await verifyPhantomReader(
+      thingSchema,
+      'from Thing access {a, b, d}',
+      '[Thing]',
+      '[Thing {a: Text, b: Text, d: Text}]'
+    );
+    await verifyPhantomReader(
+      thingSchema,
+      'from Thing access {a, b, d}',
+      '&Thing',
+      '&Thing {a: Text, b: Text, d: Text}'
+    );
+    await verifyPhantomReader(
+      thingSchema,
+      'from Thing access {a, b, d}',
+      '[&Thing]',
+      '[&Thing {a: Text, b: Text, d: Text}]'
+    );
+    const thingSchemaWithLists = `
+      schema Thing
+        a: List<Text>
+        b: List<Text>
+        c: List<Text>
+        d: List<Text>
+        e: List<Text>
+    `;
+    await verifyPhantomReader(
+      thingSchemaWithLists,
+      'from Thing access {a, b, d}',
+      'Thing',
+      'Thing {a: List<Text>, b: List<Text>, d: List<Text>}'
+    );
+    await verifyPhantomReader(
+      thingSchemaWithLists,
+      'from Thing access {a, d}',
+      'Thing',
+      'Thing {a: List<Text>, d: List<Text>}'
+    );
+    // Inline entity tests
+    const thingSchemaWithInline = `
+      schema Thing
+        foo: inline Foo {
+          a: Text,
+          b: Text,
+          c: Text,
+          d: Text,
+          e: Text
+        }
+        bar: Text
+    `;
+    await verifyPhantomReader(
+      thingSchemaWithInline,
+      `from Thing access {
+         foo {a, b, d}
+      }`,
+      'Thing',
+      'Thing {foo: inline Foo {a: Text, b: Text, d: Text}}'
+    );
+    await verifyPhantomReader(
+      thingSchemaWithInline,
+      `from Thing access {
+         foo {a, d},
+         bar
+      }`,
+      'Thing',
+      'Thing {foo: inline Foo {a: Text, d: Text}, bar: Text}'
+    );
+    // List of inline entity tests
+    const thingSchemaWithInlineList = `
+      schema Thing
+        foo: List<inline Foo { a: Text, b: Text, c: Text, d: Text, e: Text }>
+        bar: Text
+    `;
+    await verifyPhantomReader(
+      thingSchemaWithInlineList,
+      `from Thing access {
+         foo {a, d, e},
+      }`,
+      'Thing',
+      'Thing {foo: List<inline Foo {a: Text, d: Text, e: Text}>}'
+    );
+    await verifyPhantomReader(
+      thingSchemaWithInlineList,
+      `from Thing access {
+         foo {a, d},
+         bar
+      }`,
+      'Thing',
+      'Thing {foo: List<inline Foo {a: Text, d: Text}>, bar: Text}'
+    );
+    const thingSchemaWithTuples = `
+      schema Thing
+        a: (Text, Number)
```

Comment:
Ah, tuples, when will they be supported E2E? You're forward looking 🥇

Reply:
:)